### DISCUSSION TITLE
Select the server with PIA

### DISCUSSION THREAD

**Discussion by ethgrim** [haugene/docker-transmission-openvpn#2408] [Q&A] [answered]
Hello PIA seems very slow (especially in upload) but is seems that some server are better than others.
Is it possible to select the server I prefer ?

**Comment by haugene** [comment, marked as answer]
Hey :wave: Yes, this is possible. You can use the `OPENVPN_CONFIG` environment variable for this. It is described in the config section of the documentation: https://haugene.github.io/docker-transmission-openvpn/config-options/#network_configuration_options

The valid config options are not listed anywhere because we actually pull the latest configs from PIA when you start the container, so we don't know the names up-front. But if you try to start it with an invalid config name it will list the ones it found so you can choose one of them.

For example, I just ran: `docker run --rm -it -e OPENVPN_PROVIDER=pia -e OPENVPN_USERNAME=dummy -e OPENVPN_PASSWORD=dummy -e OPENVPN_CONFIG=dummy haugene/transmission-openvpn:edge` and the logs end like this:
```
Extract OpenVPN config bundle into PIA directory /etc/openvpn/pia
Supplied config dummy.ovpn could not be found.
Your options for this provider are:
albania.ovpn
algeria.ovpn
andorra.ovpn
argentina.ovpn
armenia.ovpn
au_melbourne.ovpn
au_perth.ovpn
au_sydney.ovpn
austria.ovpn
bahamas.ovpn
bangladesh.ovpn
belgium.ovpn
brazil.ovpn
bulgaria.ovpn
ca_montreal.ovpn
ca_ontario.ovpn
ca_toronto.ovpn
ca_vancouver.ovpn
cambodia.ovpn
china.ovpn
cyprus.ovpn
czech_republic.ovpn
de_berlin.ovpn
de_frankfurt.ovpn
default.ovpn
denmark.ovpn
egypt.ovpn
estonia.ovpn
finland.ovpn
france.ovpn
georgia.ovpn
greece.ovpn
greenland.ovpn
hong_kong.ovpn
hungary.ovpn
iceland.ovpn
india.ovpn
ireland.ovpn
isle_of_man.ovpn
israel.ovpn
italy.ovpn
japan.ovpn
kazakhstan.ovpn
latvia.ovpn
liechtenstein.ovpn
lithuania.ovpn
luxembourg.ovpn
macao.ovpn
macedonia.ovpn
malta.ovpn
mexico.ovpn
moldova.ovpn
monaco.ovpn
mongolia.ovpn
montenegro.ovpn
morocco.ovpn
netherlands.ovpn
new_zealand.ovpn
nigeria.ovpn
norway.ovpn
panama.ovpn
philippines.ovpn
poland.ovpn
portugal.ovpn
qatar.ovpn
romania.ovpn
saudi_arabia.ovpn
serbia.ovpn
singapore.ovpn
slovakia.ovpn
south_africa.ovpn
spain.ovpn
sri_lanka.ovpn
sweden.ovpn
switzerland.ovpn
taiwan.ovpn
turkey.ovpn
uk_london.ovpn
uk_manchester.ovpn
uk_southampton.ovpn
ukraine.ovpn
united_arab_emirates.ovpn
us_atlanta.ovpn
us_california.ovpn
us_chicago.ovpn
us_denver.ovpn
us_east.ovpn
us_florida.ovpn
us_houston.ovpn
us_las_vegas.ovpn
us_new_york.ovpn
us_seattle.ovpn
us_silicon_valley.ovpn
us_texas.ovpn
us_washington_dc.ovpn
us_west.ovpn
venezuela.ovpn
vietnam.ovpn
NB: Remember to not specify .ovpn as part of the config name.
```

You can also list multiple servers and one will be randomly selected on startup. Ie: `OPENVPN_CONFIG=sweden,spain,norway`

**Reply by ethgrim** [reply]
Thanks for your help, I'll try it later.
I still have 2 questions:
- If I don't specify anything, how the server will be selected ?
- Is there a way to select the fastest server for download and upload my torrent ? (because my problem is uploading is very slow and clunky).

Thanks again

**Reply by pkishino** [reply]
if you don't specify anything, default.ovpn is used but I don:t know which server that is linked to..
No, currently there is no way to select fastest as there is no connection test or anything.. if someone were to create such a script for PIA then that could be added

**Reply by haugene** [reply]
The default seems to be chosen at random after the configs are downloaded and extracted. https://github.com/haugene/docker-transmission-openvpn/blob/master/openvpn/pia/configure-openvpn.sh#L31-L32

Some providers have an API endpoint where you can get stats on how busy each server is, that way you can try to connect to one that is less busy. But "fastest" could be more relative I guess. Either way we don't have any mechanism to connect and disconnect from many servers and do a speed test them.

You could do it yourself I guess. [Mount a script](https://haugene.github.io/docker-transmission-openvpn/config-options/#custom_scripts) to transmission-pre-start for example, run a speed test and signal openvpn to shut down if the speed isn't satisfactory. Then you can list multiple servers in your config and it will round-robin until it finds one that works faster.

But the slowness could come at any point I would imagine, so you might need a process that periodically checks the speed. It's outside the scope of what we're trying to do here, for now at least. If you find a very elegant solution we welcome PRs and then we can see :smile:
